### ISSUE TITLE
Need some assistance with memory issue (configuration?)

### ISSUE THREAD

**Issue by achin84** [graphile/postgraphile#851]
I'm submitting a ... 

* [ ] bug report
* [ ] feature request
* [x] question 

Postgres version:
psql (10.5 (Ubuntu 10.5-1.pgdg18.04+1))

Postgraphile version:
4.0.0-rc.3

Node version:
v10.8.0

Hi all,

I've been working on some code to communicate data to my postgres server through postgraphile, but I keep running into memory issues with the postgraphile process.  I've reduced my code/database down to a very simple project, but I'm still running into issues - As I am new to postgraphile, I suspect there is some configuration or the like which I'm omitting.

Here's the process I've followed so far (I'll be using placeholders for some of the database fields/other project specific stuff).

The postgres db consists of 2-3 tables, of which there are no constraints, depenendencies, or indexes (or anything else really - just a dozen or so fields on each).

First I run the postgres/postgraphile:

> sudo postgraphile -c "postgres://root:password@localhost/test_db"

Next I run some java code which loops and posts a bunch of mutations to the postgraphql endpoint.
```
public static void main(String[] args) {
	for(int x = 0; x<1000000; x++)
    	{
    		try {
				Thread.sleep(10);
			} catch (InterruptedException e1) {
				// TODO Auto-generated catch block
				e1.printStackTrace();
			}
    		try {
    			
    			Random rng = new Random();
        		String content      = "{\"query\":\"mutation{ createTest(input: {createTest: {param1:"+rng.nextInt(100)+
        				",param2:"+rng.nextInt(100)+",param3:-1,param4:"+rng.nextInt(100)+",param5:0,param6:"+rng.nextInt(100)+
        				",param7:"+rng.nextInt(100)+",param8:"+rng.nextInt(100)+
        				",param9:935,param10:1,param11:0,param12:"+rng.nextInt(100)+
        				",param13:-1,param14:-30000,param15:5,param16:"+rng.nextInt(100)+
        				",param17:\\\"345345\\\",param18:3,param19:12,param20:\\\"Fri Apr 20 13:58:24 PDT 2018\\\"}}) {clientMutationId}  }\"}";
    		    

    			URL url = new URL("http://127.0.0.1:5000/graphql");
    			HttpURLConnection conn = (HttpURLConnection) url.openConnection();
    			conn.setDoOutput(true);
    			conn.setRequestMethod("POST");
    			conn.setRequestProperty("Content-Type", "application/json");

    			OutputStream os = conn.getOutputStream();
    			os.write(content.getBytes());
    			os.flush();
    			
    			BufferedReader br = new BufferedReader(new InputStreamReader(
    					(conn.getInputStream())));

    			String output;
    			System.out.println("Output from Server .... \n");
    			while ((output = br.readLine()) != null) {
    				System.out.println(output);
    			}

    			conn.disconnect();

    		  } catch (MalformedURLException e) {

    			e.printStackTrace();

    		  } catch (IOException e) {

    			e.printStackTrace();

    		 }
    	}
}
```
After around 100k-200k inserts, I run out of memory on the process.  The error message I get is:

>  JS stacktrace
> 
> ==== JS stack trace =========================================
> 
>     0: ExitFrame [pc: 0xffaf70841bd]
>     1: StubFrame [pc: 0xffaf73ad416]
> Security context: 0x20207ab9e789 <JSObject>
>     2: nextTick [0x4bb69edf171] [internal/process/next_tick.js:1] [bytecode=0x15c01c8a6e39 offset=0](this=0x25c513683bd1 <process map = 0x29dee614f259>,callback=0x7451922b3d1 <JSFunction maybeReadMore_ (sfi = 0x2731881b2311)>)
>     3: arguments adaptor frame: 3->1
>     4: onread [0x189ea9629ab1] [net.js:~621] [pc=0xffaf73d42...
> 
> FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
>  1: 0x8b8210 node::Abort() [node]
>  2: 0x8b825c  [node]
>  3: 0xac1d1e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
>  4: 0xac1f38 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
>  5: 0xeb11f2  [node]
>  6: 0xebd14a v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
>  7: 0xebdab4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
>  8: 0xec03e5 v8::internal::Heap::AllocateRawWithRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
>  9: 0xe888c4 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
> 10: 0x112a2ae v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
> 11: 0xffaf70841bd 
> Aborted (core dumped)

I've tried halting the process part way through, but the memory on the process doesn't seem to really free up at all either - I wasn't sure if the node process should be freeing memory periodically?

Thanks for the help, and apologies if this is a dumb problem(which it likely is) that has already been answered.  I poked around a bit but I had trouble finding an answer.

-Andy

**Comment by sjmcdowall** [comment]
Is this reproducible in -rc.4 ?? That's the latest one ..

**Comment by benjie** [comment]
I'm going to guess this is the query cache causing issues since you're using lots of dynamic queries rather than queries with variables; please create a file `.postgraphilerc.js` in the folder that you're running `postgraphile` from and make the contents of it this:

```js
module.exports = {
  options: {
    queryCacheMaxSize: 0,
  },
};
```

Then try again - please let me know whether or not it fixes it.

(If it is this that's the problem then it's definitely a bug and not anything you're doing wrong - sorry!)

**Comment by benjie** [comment]
Correction, set it to `1` not `0`.

I can exhaust memory using this script `exhaust.sh`:

```bash
#!/bin/bash
set -e
for I in `seq 1 100000`; do
  echo "{ a1_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a2_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a3_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a4_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a5_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a6_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a7_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a8_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a9_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a10_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a11_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a12_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a13_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a14_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a15_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a16_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a17_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a18_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a19_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a20_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a21_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a22_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a23_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a24_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a25_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a26_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a27_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a28_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a29_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a30_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a31_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a32_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a33_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a34_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a35_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a36_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a37_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a38_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a39_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a40_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a41_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a42_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a43_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a44_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a45_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a46_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a47_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a48_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a49_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a50_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a51_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a52_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a53_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a54_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a55_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } a56_$1$I: query { nodeId $1$I: query { nodeId $1$I: query { $1$I: nodeId } } } nodeId }" | curl -s -XPOST -d @- -H 'Content-Type: application/graphql' http://localhost:5000/graphql >/dev/null;
  /bin/echo -n '.';
done
```

combined with this for expediency:

```bash
(for ltr in aa bb cc dd ee ff gg hh ii jj kk ll mm nn oo pp qq rr ss tt uu vv ww xx yy zz; do echo $ltr; done) | parallel ./exhaust.sh
```

Climbs to over a gig of RAM in a short period of time.

Using the (corrected) workaround above memory does not exhaust.

**Comment by achin84** [comment]
@benjie Thanks - I'll try that out.  Was testing with the 0, and it does look like the memory usage is approaching the crash (slowly).  

I just wanted to verify that I added the .postgraphilerc.js file in the correct location - I tossed it in my home directory (where I've been calling the sudo postgraphile -c "postgres://root:password@localhost/test_db" from, as well as the /usr/local/bin directory (which appears to be where the postgraphile is running from).

Either way, I suspect I have a correct file in one of the proper locations now :).  I'll kick another run off and see how it performs.

@sjmcdowall I tried using the updated version, but that didn't seem to resolve the issue.  It might have run out of memory slower though?  Not really sure.

Thanks for the help guys!
-Andy

**Comment by benjie** [comment]
Yeah, turns out that setting it to `0` makes no difference; you have to set it to `1` for some reason.

The `.postgraphilerc.js` needs to be in your current working directory - it's sourced via `${process.cwd()}/.postgraphilerc.js`.

I've experimented and found that a factor of 110 makes it work for the example I have above - it now takes almost exactly 50MB more RAM at load than it does with caching disabled.

**Comment by benjie** [comment]
This should fix it: https://github.com/graphile/postgraphile/pull/852

**Comment by achin84** [comment]
Awesome - thanks!

I tested out with my dummy code, and it didn't run the memory up at all.  I also tried with my real code, and it did a lot better than before (got through about 1.5M inserts out of 1.8M, instead of failing around 172,000).  I don't have any waiting in that software,so I suspect it'll be fine with the memory once I slow down the rate I'm throwing the mutations through (That one is sending bulk mutations at 300 per packet via MQTT).

Thanks for all the help!
-Andy
